Implement PulseQueue (a consumer)

djmitche opened the pull request:
Let me know what you think.  This would replace PulseListener, but I chose "PulseQueue" to avoid confusion with that name.

As with the bare Client, the idea here is to eat all of the non-fatal errors and only report "real" errors to sentry.  So this should be a lot quieter than our existing implementations.  It's also got some facilities to avoid pounding servers during a downtime, but I'm open to suggestions for more.

This supports [unbindQueue](https://bugzilla.mozilla.org/show_bug.cgi?id=1080281) by way of `withConnection`, rather than trying to wrap that particular method.

The tests are pretty thorough, but I still haven't taken this for a test drive in a production setting, so there may be serious bugs remaining.

djmitche committed:
reorganize before pulling in a bunch more code

djmitche committed:
Expose client.namespce

djmitche committed:
Define PulseQueue for consuming messages

djmitche committed:
Include a monitor in the client

djmitche committed:
add Client.withConnection, withChannel, and activeConnection

These are the gateway methods to one-off operations like manually
binding/unbinding queues, pre-declaring excahnges, etc.

djmitche committed:
Define a minReconnectionInterval to avoid pounding servers when errors occur

jonasfj commented on README.md:
```diff
@@ -126,6 +189,76 @@ client.on('connected', async (conn) => {
+  bindings: [{           // exchange/routingKey patterns to bind to
+    exchange,            // Exchange to bind
+    routingKeyPattern,   // Routing key as string
+    routingKeyReference, // Reference used to parse routing keys (optional)
```
Unrelated, but maybe we should rename these to `pattern` and `reference` instead of `routingKeyPattern`...

We can just leave `routingKeyPattern` supported, but not documented that way existing code will keep working.

jonasfj commented on README.md:
```diff
@@ -126,6 +189,76 @@ client.on('connected', async (conn) => {
+    routingKeyPattern,   // Routing key as string
+    routingKeyReference, // Reference used to parse routing keys (optional)
+  }, ..],
+  handleMessage,         // handler for incoming messages
```
Optional, right?

djmitche replied:
No, what would we do with messages if not listening for them?

jonasfj commented on src/consumer.js:
```diff
@@ -0,0 +1,229 @@
+   * server.  This method will return after any pending consumers have
+   * completed (ack or nack).
+   */
+  async stop() {
```
Could we rename this `close`, we also need:
 * `close` to close this PulseQueue object, that is: stop listening and close the channel used by this `PulseQueue`.
 * `destroy` to delete the named queue, only supported if `queueName` was given (otherwise equivalent to `close`).

We could also add:
 * `pause` to pause message arrival
 * `resume` to resume message arrival
 * `bind` to add extra binding
 * `on('message, handler)` would also be neat to have, as this is the existing API.

To be fair I do see the appeal in just declaring your bindings and handler upfront, rather than manipulating them after the object is created. There is a lot less state to manage here. And you won't have to call `resume`.

But isn't it almost better to just do a function called `PulseConsume(...)` that returns a function I can call to close the consumer... or something like that? That way you can do async stuff in the constructor.

**Note**, I don't have a perfect design, I'm open to changing it. This design have more major flaw: I can't start listening before I create a task and observe the message, because the constructor is synchronous, so after the `PulseQueue` is created I'm still not sure I'll receive all messages, as declaration of queue and binding could happen at any time.

jonasfj replied:
Other problem I see is that you get the impression that bindings disappear, when in fact they are stateful if you use pass `queueName`

djmitche replied:
I don't see a point in `pause` and `resume` -- this starts consuming immediately, and stops  (once) on "stop".  If you want to start again, just make a new PulseQueue.

For cases beyond the common one-queue-bound-forever case, `bind` and `unbind` are done via `withChannel`.  I wrote a wrapper, but it really added no value so I deleted it.

The problem with `on('message', handler)` is what to do with messages that arrive before it's called.  You're right, it's a lot simper to do the state management when the whole listening process is defined up-front.  In practice, I don't think the two interfaces are much different for the user.

I see your point about the race condition.  Hmm.

jonasfj commented on src/consumer.js:
```diff
@@ -0,0 +1,229 @@
+        }
+      });
+      this.consumerTag = consumer.consumerTag;
+
```
How about `channel.on('error'` ? I'm not sure... just wondering if we're missing something.

djmitche replied:
Good point

jonasfj commented on src/consumer.js:
```diff
@@ -0,0 +1,229 @@
+      conn.on('retiring', () => this._shutdown());
+    } catch (err) {
+      this.client.monitor.reportError(err);
+      conn.failed();
```
So binding to an exchange that doesn't exist breaks the entire connection, and not just the channel?

djmitche replied:
Well, it causes the connection to be restarted.  There's no harm in that (and it's not a common case anyway).

jonasfj commented on src/consumer.js:
```diff
@@ -0,0 +1,229 @@
+      if (!this.running) {
+        return;
+      }
+
```
You can automatically reconnect if no `queueName` have been passed.

Doing so mean I loose guarantees. If `queueName` is not given, `reconnect` should be an option that defaults to `false`.

djmitche replied:
I think this library is basically not useful for exclusive queues.  I'll just remove support for that entirely.

djmitche commented:
The use-case you mention, listening for events about a specific task, is not uncommon.  The thing is, if you use an exclusive / auto-delete queue for that, you're already doing it wrong: you're assuming your connection will remain up until that task is resolved.  So we have two options:
 * disallow this and require use (and later deletion) of a permanent queue; or
 * allow this in such a way that the consequences are clear (your script might hang forever..)
Which do you think is better?

djmitche commented:
I think the important point here is that in the real world, there is no option to not reconnect -- connections will break, and setting `{reconnect: false}` won't change that.  So in the case of an exclusive queue, I'm not sure what the right action is when the connection fails -- `process.exit(1)`?

djmitche commented:
I'm not sure your API design is very well-thought-out.  If bind is an operation performed in a second method call:

```javascript
let q = new PulseQueue(..);
q.bind('exch', '#');
```
then does that occur once, or on every reconnection?  If it's on every reconnection, then I think it should be part of PulseQueue's initial state.

Basically, I think PulseQueue either needs to be a small wrapper to call `consume` on a queue -- and let the user handle all of the assertions of queues and whatnot, including whether they are exclusive -- or declare them for the user.  Trying to do both is not going to work.

I see that you don't like my API design, but I don't think your API design works :)

djmitche commented:
Per discussion this afternoon:
 * Continue supporting exclusive queues, but with an explicit option to the constructor
   * When an *exclusive* queue reconnects, the PulseQueue will emit an error (something like `code='ExclusiveQueueReconnection'` with a useful error message).  Users can either handle that error, or ignore it and let it crash the app.
   * We'll invent some convenient DI technique to support exclusive queues in tests (to go along with the mock PulseQueue for use without credentials)
 * The race condition can be solved with an async `start` method that does not return until the queue is asserted.
 * For cases more complicated than listen-to-this-queue-forever and exclusive queues, `client.withChannel` is the escape hatch.  If you want to delete a queue, `await client.withChannel(chan => chan.deleteQueue(queueName))`.  If you want to add or remove bindings (as tc-hooks may..) same answer.

djmitche commented:
Updated -- have a look?

djmitche committed:
make the PulseQueue constructor async

djmitche committed:
Remove the client.start method, too

This means that the client starts up immediately on creation.  It will
connect and set up a timer to reconnect, thus causing Node not to exit
if not handled correctly.  This also sets up a bit of a race to call
`client.on('connection', ..)`, but that race really only exists in the
tests.

djmitche commented:
I removed the start methods from both the Client and the PulseQueue.  Both start automatically on creation now (which necessitated PulseQueue being constructed asynchronously).

jonasfj commented on README.md:
```diff
@@ -122,10 +185,108 @@ client.on('connected', async (conn) => {
+  }, ..],
+  handleMessage,         // handler for incoming messages
+  queueName,             // Queue name; if omitted, set exclusiveQueue: true
+  exclusiveQueue,        // If true, use an exclusive queue (and lose messages on reconnect)
```
I don't like this... mostly due to ergonomics...
I'll end up writing:
```js
pulse.consume(client, {
  bindings: [...bindings],
  queueName: config.pulse.queueName,
  exclusiveQueue: config.pulse.queueName === '',
  prefetch: config.pulse.prefetch,
  handleMessage: async (message) => {
    console.log(message);
  },
});
```
Or I'll write:
```js
pulse.consume(client, {
  bindings: [...bindings],
  queueName: config.pulse.queueName,
  exclusiveQueue: profile !== 'production',
  prefetch: config.pulse.prefetch,
  handleMessage: async (message) => {
    console.log(message);
  },
});
```

When what I really want to write is more in line with:
```js
pulse.consume(client, {
  bindings: [...bindings],
  queueName: config.pulse.queueName,
  prefetch: config.pulse.prefetch,
}, async (message) => {
    console.log(message);
});
```
or
```js
client.consume({
  bindings: [...bindings],
  queueName: config.pulse.queueName,
  prefetch: config.pulse.prefetch,
}, async (message) => {
    console.log(message);
});
```
or

```js
client.consume([...bindings], {
  queueName: config.pulse.queueName,
  prefetch: config.pulse.prefetch,
}, async (message) => {
    console.log(message);
});
```

and maybe then we could even omit `options` and do:

```js
client.consume([...bindings], async (message) => {
    console.log(message);
});
```

---
I know I know... this is just ergonomics... you don't have follow what I suggest here... maybe making `handleMessage` positional makes sense as it's required, and writing it last is really pretty (and reduces indentation required). Just giving you some suggestions..

djmitche replied:
I don't really understand what these snippets are suggesting -- you wrote five different things.  I can make `handleMessage` positional easily enough.  For the rest, I think most of our APIs either take options or positional args, and that mixing the two just makes it hard to remember what's what.

jonasfj replied:
Up to you... It's all ergonomics, I just wanted to give some ideas..

Since some args are required and others aren't that might be neat.. especially if it reduces indentation..

Biboswan replied:
w.r.t to your suggestion i feel making handlemessage to be made positional and such as this
`new PulseQueue(client, options, handlemessage)`

jonasfj commented on README.md:
```diff
@@ -122,10 +185,108 @@ client.on('connected', async (conn) => {
+If `queueName` is specified, this will create a durable queue using a
+pulse-compatible queue name based on `queueName` (prefixed with
+`queue/<namespace>`).  If no `queueName` is given, `exclusiveQueue` must be
+true.
```
Using `queueName: "my-queue` with `exclusiveQueue: true`  is bad... as it:
1. makes no sense to name the queue, and,
2. creates an opportunity for bugs.

jonasfj replied:
I like being able to do `queueName: ''` and get an exclusive queue with a random name.

But if you really want to make those semantics, which will make testing very very very hard.
I would suggest two different methods `pulse.consume(client, queueName, options, handler)` and `pulse.consumeExclusive(client, options, handler)`

Or make `queueName` positional, and say that if you don't get a string passed then it's an exclusive queue... 
`pulse.consume(client, queueName, bindings, options, handler)` since `queueName` is the only string it's easy to see if it's omitted.

djmitche replied:
Testing is going to be done with mocks, anyway.

I don't want to make this library *too* opinionated.  If users want to do something we have not imagined, then getting AssertionError: we don't think you wanted to do that is not really useful.

A named exclusive/autodelete queue makes sense in some cases.  It gets you assurance that there is only one consumer, and that the queue will not fill up when not in use.  This would be good for a command-line tool, for example.

It's worth noting that we're abusing the phrase "exclusive" here to mean "exclusive and autoDelete".  So there's lots of potential sharp edges.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+      this.namespace = username;
+    }
+
+    assert(monitor, 'monitor is required');
```
Let's make `monitor` optional

djmitche replied:
It's not optional for any of our other APIs..

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+    this.lastConnectionTime = 0;
+
+    this.id = ++clientCounter;
+    this.debug = debug(`taskcluster-lib-pulse:client:${this.id}`);
```
Can we avoid this? Few apps will have multiple `client` objects... and it makes debug logs more noisy...

Biboswan replied:
so do you mean to avoid debugs completely

djmitche replied:
I think having a shorter debug prefix -- without the client ID, and probably a shorter name.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+
+  /**
+   * Get a full object name, following the Pulse security model,
+   * `<kind>/<namespace>/<name`.  This is useful for manipulating these objects
```
`<name>`

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+   * `<kind>/<namespace>/<name`.  This is useful for manipulating these objects
+   * directly, for example to modify the bindings of an active queue.
+   */
+  fullObjectName(kind, name) {
```
please consider prefixing this `_`, it feels internal...

djmitche replied:
It's a part of the Pulse conventions, so I think it's public..

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+        this.emit('connected', newConn);
+      });
+      newConn.once('finished', () => {
+        this.connections = this.connections.filter(conn => conn.id !== newConn.id);
```
Just do reference equality, that should be right?

jonasfj replied:
`this.connections = this.connections.filter(c => c !== newConn);`

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+   * the client is already connected.  This does *not* automatically re-run
+   * the function if the connection fails.
+   */
+  withConnection(fn) {
```
Why not just make a `connection()` that returns active connection or promise for a connection that is ready...

djmitche replied:
For parallelism with withChannel

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+    this.id = id;
+    this.amqp = null;
+
+    this.debug = debug(`taskcluster-lib-pulse:connection:${client.id}.${id}`);
```
I single per-connection ID makes sense...

Biboswan replied:
yeah further i prefer to have a connection id as a combo of both client id and id better for debugging

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+    this.state = 'waiting';
+  }
+
+  connect() {
```
Just make this `async` instead of having `.then`

jonasfj commented on src/consumer.js:
```diff
@@ -0,0 +1,265 @@
+
+      // consider any errors on the channel to be potentially fatal to the
+      // connection (better safe than sorry)
+      channel.on('error', () => conn.failed());
```
No, channel errors are not connection errors.

RabbitMQ is not insane :)

If we have an error at channel, the channel is dead, typically because you made a mistake... But typically, it won't kill the connection, if it does having `.on('error', ` on the connection should be sufficient.

djmitche replied:
I added this at your suggestion, iirc :)

djmitche replied:
So the alternative here may be to repeat the consumer's `createChannel` using the current amqp connection.  In so doing, we need to make sure that we don't set up *two* channels at the same time (once because of a channel error, and the other because of a connection error or periodic recycle).  So we'll need some kind of locking.

I think this is the simpler solution, and given the efforts taken to ensure that a recycle has no application impact, I think it's OK.

Note that you do get an error thrown when you call `channel.someMethod(..)` in an inappropriate way.  So we're not ignoring errors here..

Biboswan replied:
differentiating between connection and channel errors by experimenting will make things clear.If channel error is due to some user's invalid data through `channel.someMethod` then repeating `createChannel` may not be right or there should be no. of reattempt limits.

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+      });
+  }
+
+  failed() {
```
I think this is wrong. connection should only be considered dead if it throws `on('error', ..)`

Biboswan replied:
when amqp throws error, failed() is called. Do u want something like `connection.emit('error',failed)`

djmitche replied:
Recycling a connection shouldn't be a big deal.  The idea here is to exercise an abundance of caution, rather than relying only on the `error` event at the `connection` level.  Whether to consider any error from a channel as fatal to the connection is debatable.

Biboswan replied:
i think  if its only a channel error it may not be fatal to the connection, suppose `source`(exchange) is invalid for `channel.bindQueue` then  if that's only a channel error then that channel should be closed but not entire connection and recycled.

djmitche replied:
Does that produce a channel error?  We need to understand better when the channel emits an `error` event, and when it throws an error from a method.  Does it always do both?

Failing to handle error events results in termination of the process, so we need to handle them *somehow*.  Whether that's ignoring them (because the caller of some method got an error) or recycling the connection, or trying to re-create the channel (but carefully to not create two channels at the same time), I'm not sure.

Perhaps some experimentation with the API is in order:
 * Call a method with the wrong arguments
 * Call a method on a channel attached to a closed connection (for example, after killing the rabbitmq process)

Biboswan replied:
I ran a cloudamqp server and observed that

- an error in channel closes that channel but other channels in the connection are not harmed. That is to say no connection close or error event occurs for that.

Further documentation says, well should have been seen that part before

Connection errors :
- a protocol transgression the server detected (likely a bug in this library)
- a server error
- a network error
- the server thinks the client is dead due to a missed heartbeat

Channel errors:
- an operation failed due to a failed precondition (usually something named in an argument not existing)
- a human closed the channel with an admin tool
Further a channel will not emit 'error' if its connection closes with an error.

Logically, if the connection close event occurs, first channels are closed followed by the connection...

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+    this.emit('retiring');
+
+    // actually close this connection 30 seconds later
+    setTimeout(() => {
```
`await new Promise(resolve => setTimeout(resolve, this.client.retirementDelay))`

jonasfj commented on src/client.js:
```diff
@@ -0,0 +1,351 @@
+
+    this.recycleInterval = recycleInterval || 3600 * 1000;
+    this.retirementDelay = retirementDelay || 30 * 1000;
+    this.minReconnectionInterval = minReconnectionInterval || 15 * 1000;
```
I think that these should be made private with a `_` so people don't unwilling read these properties..

this goes for a lot of the properties and methods in all classes... If things are not intented to be public prefixing `_` is a good way to go.

jonasfj reviewed (dismissed):
ergonomics is a small part here... we can also ignore that for now...

djmitche commented:
I think I'm going to drop this.  It's turning out to be a much longer project than I thought, and it's not related to redeployability.

djmitche commented:
@Biboswan is going to take over work on this PR, starting with the review comments.

Biboswan committed:
Made connect method async

Biboswan committed:
recycleInterval,retirementDelay, minReconnectionInterval made private

Biboswan committed:
Connection start closing in retire  after 30 secs

Biboswan committed:
Fix typo

Biboswan committed:
Fix minor equality logic

Biboswan commented:
https://github.com/taskcluster/taskcluster-lib-pulse/blob/consumer/src/client.js#L216
is this line necessary? i think when the connection is dead, `close` event handler automatically calls `recycle` method

djmitche commented:
Yes, that's a good point.

Biboswan committed:
Catch error on create channel failure

Biboswan committed:
Filter invalid binding and start binding process again

Biboswan committed:
Add test consume messages with some incorrect bindings

Biboswan commented:
I think `createQueue` and `bindqueue` to be seperate functions as it would reduce connection attemps when we start the binding process again after filtering invalid binding. We dont need to try to createQueue once already done.

Biboswan commented:
@jonasfj have a look

djmitche commented:
Rebased onto master (the changes on master were minor).

I'm stepping back in to work on this now.

Our uses of this library will be in building microservices, which listen on durable queues.  So:
 * We don't need to support auto-delete / exclusive queues -- if we later want to make a new "high level" class to support such, that's OK.
 * Being conservative in error handling (recycling the connection when, perhaps, we only need to recycle the channel) is OK as it errs on the side of keeping services running.

djmitche commented:
Issues to address:
 * [x] rename to PulseConsumer
 * [x] verify all semantics (sync/async) are represented in README
 * [x] README: whitespace in "Manipulating AMQP Objects"
 * [x] README: "for each messages"
 * [x] README: withChannel / withConnection take async funcs
 * [x] README: remove redundant "Listening starts immediately, or when the client starts."
 * [x] README: `objectName` -> `fullObjectName`
 * [x] make `handleMessage` positional
 * [x] drop `exclusiveQueue` support
 * [x] revert withChannel changes (flag, returntype)
 * [x] revert making retire async
 * [x] revert `_createAndBindQueue` flags
 * [x] call channel.on(error) (already done)
 * [x] shorter debug lines
 * [x] wrapper for conn.on('connected') that calls if it's already connected

djmitche committed:
Revert "Filter invalid binding and start binding process again"

This reverts commit 17d89ad81bab430413a20224d3ea614b6f226ecb.

djmitche committed:
Revert "Catch error on create channel failure"

This reverts commit 2795beb5a1499c45b8e360c6e725dda7c05dbf23.

djmitche committed:
Revert "Connection start closing in retire  after 30 secs"

This reverts commit ff89cb7f3892cd6b08eb707cd5621eec299b984c.

djmitche commented:
I've reverted ff89cb7f3892cd6b08eb707cd5621eec299b984c, 2795beb5a1499c45b8e360c6e725dda7c05dbf23, and 17d89ad81bab430413a20224d3ea614b6f226ecb.  I can't see the reason for them, and I think they will cause the library to ignore errors (in ff89cb7f3892cd6b08eb707cd5621eec299b984c by not binding to requested exchanges and in 2795beb5a1499c45b8e360c6e725dda7c05dbf23 by returning successfully from `withChannel` without ever running `fn`).  I'm also not keen on the use of integers in a variable named `flag` to indicate runtime conditions -- named variables are a better choice.

I don't think there's a reason to make retirement async, either -- the function just indicates that the current connection should be retired, which is an instantaneous thing.  It doesn't need to wait for that to complete.  17d89ad81bab430413a20224d3ea614b6f226ecb also had some bugs: there are calls to `retire()` without `await`, and the README is not updated to reflect that the function is now async.

@Biboswan if I've missed something here, let me know what it is and maybe we can make some adjustments.

djmitche committed:
Revert "Add test consume messages with some incorrect bindings"

This reverts commit 4f3ebe3ee50da3118ec86f74650814e3d52e8665.

djmitche committed:
rename PulseQueue to PulseConsumer

djmitche committed:
Minor README updates

djmitche committed:
Make consumer's handleMessage a positional argument

djmitche committed:
drop support for ephemeral queues in PulseConsumer

djmitche committed:
shorten debug prefixes

djmitche committed:
add an onConnected method to help users avoid races

djmitche committed:
remove unnecessary non-dev dependency

Biboswan commented:
@djmitche I shall look at once and let you know if there's anything missing

imbstack commented on src/client.js:
```diff
@@ -59,43 +60,51 @@ exports.buildConnectionString = buildConnectionString;
       assert(!password, 'Can\'t use `password` along with `connectionString`');
       assert(!hostname, 'Can\'t use `hostname` along with `connectionString`');
-      assert(!vhost, 'Can\'t use `hostname` along with `connectionString`');
+      assert(!vhost, 'Can\'t use `vhost` along with `connectionString`');
```
Wow, there were a lot of little nitpicks we missed here before you got to it.

imbstack commented on src/client.js:
```diff
@@ -128,20 +137,110 @@ class Client extends events.EventEmitter {
+   * `<kind>/<namespace>/<name>`.  This is useful for manipulating these objects
+   * directly, for example to modify the bindings of an active queue.
+   */
+  fullObjectName(kind, name) {
```
Should we assert `kind` and `name` here to avoid creating something like `undefined/foo/undefined` or whatever? It probably isn't a big issue.

imbstack reviewed (approved):
I can give a more detailed review next week as well if you'd like, but it looks like jonas already put you through the wringer :p

djmitche committed:
assert that kind, name exist in fullObjectName
